proposed/resource-limit/resource-limit.md

```diff
@@ -0,0 +1,316 @@
+
+## Scenarios and User Experience
+
+Excerpt from https://microsoft.sharepoint.com/:w:/r/teams/NETRateLimitingInterface/_layouts/15/Doc.aspx?sourcedoc=%7B557C73D8-F2A2-4799-82D0-6A5C13F58E4F%7D&file=Rate%20Limiting%20interface_CBB%20Perspective.docx&nav=eyJjIjo2MDMwNjMzMzZ9&action=default&mobileredirect=true&cid=fb8aa981-06d2-4c69-9bbd-ddfb278e2eb7
+
+> Outages caused when system activities exceed the system’s capacity is a leading concern in system design.  The ability to handle system activity efficiently, and graceful limit the rate of activities executing before the system is under stress is a fundamental principle in system resiliency.  This applies not only to interactions between 2 or more disparate systems but also to the interactions occurring within all layers of an individual system instance.  ASP.NET Core does not have a standardized means for expressing and managing rate limiting logic needed to produce a resilient system.  This adds complexity to designing and developing resilient software in ASP.NET Core by introducing an easy vector for competing rate limiting logic and anti-patterns.   A standardized interface in ASP.NET Core for rate limiting activities will make it easier for developers to build resilient systems for all scales of deployment and workload.  This document outlines a standard rate limiting interface with the goal of adding it to ASP.NET Core.
+
+## Requirements
+
+### Goals
+
+Users will interact with this component in order to obtain decisions for rate or concurrency limits. This abstraction require explicit release sematics in the style of Semaphores to accommodate non self-replenishing (i.e. concurrency) resources. This component encompasses the TryAcquire/AcquireAsync mechanics (i.e. check vs wait behaviours) and default implementatinos will be provided for select accounting method (fixed window, sliding window, token bucket, simple concurrency). The return type is a `ThrottledResource` which manages the lifecycle of the aquired resources. This API should allow for a simple implementation that keeps an internal count of the underlying resource but also allow the use of an external resource count storage.
+
+For bucketized policies (for example, rate limit by IP), all bucket will be represented by one limiter. The complexity of how bucketing is computed for a request will live in the policy component which computes a resource ID.
+
+We expect that the user can opt between different implementations of the proposed interface to choose between different rate limiting algorithms. For example, between an implementation that uses token bucket vs one that makes rate limiting decisions based on ambient environment such as CPU usage. Also, different implementations may choose to use a queue/stack/priorityqueue when waiting for resources to become available.
+
+### Non-Goals
+
+TDB
+
+## Stakeholders and Reviewers
+
+- .NET team (System.Threading)
```

Comment:
Networking too? (https://github.com/dotnet/runtime/issues/52079#issuecomment-831375001)

Reply:
Also, add details about how you expect it to be used by the low-level networking stack.